crates/functional-yew/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+use std::borrow::Borrow;
+use std::cell::RefCell;
+use std::ops::DerefMut;
+use std::rc::Rc;
+use yew::{Component, ComponentLink, Html, Properties};
+
+thread_local! {
+    static CURRENT_HOOK: Rc<RefCell<Option<Rc<RefCell<HookState>>>>> = Rc::new(RefCell::new(None));
+}
+
+struct HookState {
+    counter: usize,
+    render_call: Rc<dyn Fn()>,
+    hooks: Vec<Option<Box<dyn std::any::Any>>>,
+}
+
+pub trait FunctionProvider {
+    type TProps: Properties + PartialEq;
+    fn run(props: &Self::TProps) -> Html;
+}
+
+pub struct FunctionComponent<T: FunctionProvider> {
+    _never: Option<std::marker::PhantomData<T>>,
+    props: T::TProps,
+    hook_state: RefCell<Option<Rc<RefCell<HookState>>>>,
```

Comment:
If we go with my suggested changes, we can simplify to `RefCell<Option<HookState>>`